### DISCUSSION TITLE
Pattern Matching: not-null variable extraction

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#7403] [Language Ideas]
Much criticism has been made of the pattern matching construct required to extract a variable from a not-null pattern:
```csharp
if (MyProp is { } x)
{
    // …
}
```
this isn’t intuitively understandable as “not null” and is a little clunky, and this isn’t to mention the `not` version of this pattern

While extracting a variable when it is non-null is technically a no-op, it’s extremely useful when you want to access a property only once, or make a method call return rather sleek. I fully acknowledge the above example can be re-written as 

```csharp
var x = MyProp;
if (x is not null)
{
    // …
}
```

however the more likely scenario (especially for newer developers who don’t know accessing a property multiple times can have issues) is

```csharp
if (MyProp is not null)
{
    // Use MyProp directly here
}
```

# Proposal

I propose a syntax of 

```csharp
if (MyProp is (not null) x)
{
    // …
}
```
the above code tells the full story clearly and concisely. 

This doesn’t handle the `not` scenario but that is for another proposal (`not CONST` variable extraction anyone?)

**Comment by DavidArno** [comment]
An existing alternative is:
```cs
if (MyProp is var x and not null)
{
   // ...
}
```

**Comment by En3Tho** [comment]
Why so you call 2(3)-chars pattern "clunky" and then introduce an alternative of 10 chars? How it is less "clunky"? Or how is it more concise?

**Reply by RenderMichael** [reply]
It’s not just about the character count, it’s about readability

**Reply by En3Tho** [reply]
`{}` is totally readable. And also easily expandable to a bigger pattern if needed.

**Reply by RenderMichael** [reply]
As experienced C# developers, we know exactly what `{ }` means. But that's not necessarily intuitive to new devs, devs coming from other languages, or even C# devs who work with C# <8.0, ie .NET Framework.

**Reply by En3Tho** [reply]
First of all to use both `{}` and `not null` you have to know what pattern matching is, what syntax it uses, what rules are there. It's part of the feature itself. I don't get your point.

**Comment by HaloFour** [comment]
I don't care for wrapping the pattern in parenthesis.  That definitely feels like trying to work around details of the language specification, but does so in a manner that I don't think makes sense as you're not attempting to establish a precedence.  The underlying issue is really that `null` patterns don't allow introducing a variable, and while that by itself wouldn't make sense, it does make sense when negated with a `not` pattern.  So I would argue that it would make more sense for `null` patterns to allow introducing variables.  Or for the compiler to understand the combination of `not` and `null` patterns as special.  That way `x is not null y` would just be legal without requiring parenthesis or other acrobatics.

**Reply by RenderMichael** [reply]
I added the parentheses so the pattern isn’t `x is not (null y)` which would be constant variable extraction. That might be a road to go down, for example what about 
```csharp
if (x is null y) { }
else
{
    // y is available here
}
```

**Reply by HaloFour** [reply]
> for example what about

I guess with `null` patterns that questions would be around definite assignment, which would probably have to be handled differently from most other variable patterns.  But I don't know how that would work with the `not` pattern out of the box without additional tweaks to it's behavior.

**Reply by HaloFour** [reply]
Throwing some spaghetti at the wall: #7404

**Comment by Richiban** [comment]
This would be useful in other (similar) contexts too; I found myself writing this yesterday:

```cs
if (x.CompareTo(y) is var c and not 0)
{
    return c;
}

// carry on comparing other things
```

It would be an improvement to be able to say `is (not 0) c` but it's pretty minor. I'm not 100% sure it's worth the LDT's time.
